gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java

```diff
@@ -0,0 +1,197 @@
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) {
+                List<Integer> rpcResponse = invocation.getArgument(0);
+                List<SettableApiFuture<Integer>> apiFutures = invocation.getArgument(1);
+                apiFutures.get(0).set(rpcResponse.get(0));
+                return null;
+              }
+            })
+        .when(mockDescriptor)
+        .splitResponse(Mockito.<Integer>anyList(), Mockito.<SettableApiFuture<Integer>>anyList());
+
+    ApiFuture<Integer> result = underTest.add(4);
+    underTest.flush();
+    assertThat(result.isDone()).isTrue();
+    assertThat(result.get()).isEqualTo(16);
+
+    verify(mockDescriptor, times(1)).newRequestBuilder(labeledIntList);
+    verify(mockDescriptor, times(1))
+        .splitResponse(Mockito.<Integer>anyList(), Mockito.<SettableApiFuture<Integer>>anyList());
+    verify(unaryCallable, times(1)).futureCall(any(LabeledIntList.class));
+  }
+
+  /** Verifies exception occurred at RPC is propagated to element results */
+  @Test(expected = ExecutionException.class)
+  public void testBatchingFailed() throws Exception {
+    final Exception exception = new RuntimeException();
+    when(mockDescriptor.newRequestBuilder(any(LabeledIntList.class)))
+        .thenReturn(SQUARER_BATCHING_DESC_V2.newRequestBuilder(labeledIntList));
+    when(unaryCallable.futureCall(any(LabeledIntList.class)))
+        .thenReturn(ApiFutures.<List<Integer>>immediateFailedFuture(exception));
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) {
+                List<SettableApiFuture<Integer>> apiFutures = invocation.getArgument(1);
+                apiFutures.get(0).setException(exception);
+                return null;
+              }
+            })
+        .when(mockDescriptor)
+        .splitException(any(Throwable.class), Mockito.<SettableApiFuture<Integer>>anyList());
+    ApiFuture<Integer> failedResult = underTest.add(5);
+    underTest.flush();
+    verify(mockDescriptor, times(1)).newRequestBuilder(labeledIntList);
+
+    verify(mockDescriptor, times(1))
+        .splitException(any(Throwable.class), Mockito.<SettableApiFuture<Integer>>anyList());
+    verify(unaryCallable, times(1)).futureCall(any(LabeledIntList.class));
+    assertThat(failedResult.isDone()).isTrue();
+    try {
+      failedResult.get();
+    } catch (InterruptedException | ExecutionException e) {
+      Truth.assertThat(e.getCause()).isInstanceOf(RuntimeException.class);
+      throw e;
+    }
+  }
+
+  /** Tests accumulated element are resolved when {@link Batcher#flush()} is called. */
+  @Test
+  public void testBatchingWithCallable() throws Exception {
+    underTest = newBatcherInstance();
+    int limit = 100;
+    int batch = 10;
+    List<ApiFuture<Integer>> resultList = new ArrayList<>(limit);
+    for (int i = 0; i <= limit; i++) {
```

Comment:
Why does this have to happen in a loop? it will be easier to read if you just add a single element, verify that its result is not ready, flush, verify that its ready now. And maybe do it again with a second element